Use ROCm 4.3 in CI

WeiqunZhang opened the pull request:
Since the VOP bug has been fixed in the latest ROCm release v4.3, CI can use
the latest ROCm release now instead of 4.1.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

WeiqunZhang committed:
Use ROCm 4.3 in CI

Since the VOP bug has been fixed in the latest ROCm release v4.3, CI can use
the latest ROCm release now instead of 4.1.

ax3l commented:
Interesting that it cannot find `MPI_CXX`. Trying locally now.

Fun, reason found in `CMakeError.log`:
```
/opt/rocm/llvm/bin/clang++       -pthread CMakeFiles/cmTC_abf94.dir/test_mpi.cpp.o  -o cmTC_abf94  -Wl,-rpath,/usr/lib/x86_64-linux-gnu/openmpi/lib /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so 
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_class_init_epoch [--no-allow-shlib-undefined]
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_list_item_t_class [--no-allow-shlib-undefined]
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_class_initialize [--no-allow-shlib-undefined]
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_uses_threads [--no-allow-shlib-undefined]
```

```
$ mpiexec --version
mpiexec (OpenRTE) 4.0.3
```

ax3l commented:
Looks like an Ubuntu bug, they should link the `libopen-pal.so` that is in the OpenMPI libs but they don't.

```
$ mpic++ --showme
g++ -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include -pthread -L/usr/lib/x86_64-linux-gnu/openmpi/lib -lmpi_cxx -lmpi
root@f57c056c9326:/opt/amrex# less /opt/amrex/build_nofortran/CMakeFiles/CMakeError.log
```

We can work-around it with `export LDFLAGS="-lopen-pal"`

Filed upstream bug reports:
- https://bugs.launchpad.net/ubuntu/+source/openmpi/+bug/1941786
- https://github.com/open-mpi/ompi/issues/9317

ax3l committed:
Work-Around: OpenMPI 4.0.3 on Ubuntu 20.04

"mpic++ --showme" forgets to link open-pal
```
/opt/rocm/llvm/bin/clang++       -pthread CMakeFiles/cmTC_abf94.dir/test_mpi.cpp.o  -o cmTC_abf94  -Wl,-rpath,/usr/lib/x86_64-linux-gnu/openmpi/lib /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_class_init_epoch [--no-allow-shlib-undefined]
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_list_item_t_class [--no-allow-shlib-undefined]
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_class_initialize [--no-allow-shlib-undefined]
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_uses_threads [--no-allow-shlib-undefined]
```

Results in `MPI_CXX` being not found.

ax3l reviewed (approved):
Thanks! :)

WeiqunZhang commented:
@ax3l Thanks for figuring out the openmpi issue!

ax3l commented:
Upstream update: OpenMPI 4.0.7+ will ship the fix (https://github.com/open-mpi/ompi/pull/9077).
